precompiles/randomness/src/lib.rs

```diff
@@ -0,0 +1,466 @@
+		let gas_limit = input.read::<u64>()?;
+		let salt = input.read::<H256>()?;
+		let num_words = input.read::<u8>()?;
+		handle.record_cost(
+			REQUEST_RANDOMNESS_ESTIMATED_COST + RuntimeHelper::<Runtime>::db_read_gas_cost(),
+		)?;
+		let two_epochs_later =
+			<Runtime as pallet_randomness::Config>::BabeDataGetter::get_epoch_index()
+				.checked_add(2u64)
+				.ok_or(error("Epoch Index (u64) overflowed"))?;
+		let request = Request {
+			refund_address,
+			contract_address,
+			fee,
+			gas_limit,
+			num_words,
+			salt,
+			info: RequestType::BabeEpoch(two_epochs_later),
+		};
+
+		let request_id: U256 = Pallet::<Runtime>::request_randomness(request)
+			.map_err(|e| error(alloc::format!("{:?}", e)))?
+			.into();
+
+		Ok(PrecompileOutput {
+			exit_status: ExitSucceed::Returned,
+			output: EvmDataWriter::new().write(request_id).build(),
+		})
+	}
+	/// Make request for local VRF randomness
+	fn request_local_randomness(handle: &mut impl PrecompileHandle) -> EvmResult<PrecompileOutput> {
+		let mut input = handle.read_input()?;
+		let contract_address = handle.context().caller;
+		let refund_address = input.read::<Address>()?.0;
+		let fee: BalanceOf<Runtime> = input
+			.read::<U256>()?
+			.try_into()
+			.map_err(|_| revert("amount is too large for provided balance type"))?;
+		let gas_limit = input.read::<u64>()?;
+		let salt = input.read::<H256>()?;
+		let num_words = input.read::<u8>()?;
+		let blocks_after_current = input.read::<u64>()?;
+		handle.record_cost(
+			REQUEST_RANDOMNESS_ESTIMATED_COST + RuntimeHelper::<Runtime>::db_read_gas_cost(),
+		)?;
+		let current_block_number: u64 = <frame_system::Pallet<Runtime>>::block_number()
+			.try_into()
+			.map_err(|_| revert("block number overflowed u64"))?;
+		let requested_block_number = blocks_after_current
+			.checked_add(current_block_number)
+			.ok_or(error("addition result overflowed u64"))?
+			.try_into()
+			.map_err(|_| revert("u64 addition result overflowed block number type"))?;
+		let request = Request {
+			refund_address,
+			contract_address,
+			fee,
+			gas_limit,
+			num_words,
+			salt,
+			info: RequestType::Local(requested_block_number),
+		};
+
+		let request_id: U256 = Pallet::<Runtime>::request_randomness(request)
+			.map_err(|e| error(alloc::format!("{:?}", e)))?
+			.into();
+
+		Ok(PrecompileOutput {
+			exit_status: ExitSucceed::Returned,
+			output: EvmDataWriter::new().write(request_id).build(),
+		})
+	}
+	/// Fulfill a randomness request due to be fulfilled
+	fn fulfill_request(handle: &mut impl PrecompileHandle) -> EvmResult<PrecompileOutput> {
+		let mut input = handle.read_input()?;
+		let request_id: u64 = input
+			.read::<U256>()?
+			.try_into()
+			.map_err(|_| error("Input RequestId overflows u64 type set in Pallet"))?;
+		// read all the inputs
+		let pallet_randomness::FulfillArgs {
+			request,
+			deposit,
+			randomness,
+		} = Pallet::<Runtime>::prepare_fulfillment(request_id)
+			.map_err(|e| error(alloc::format!("{:?}", e)))?;
```

Comment:
Does this work in WASM properly?

Reply:
`alloc::format` is no_std so it will work